documentation/Calls.md

```diff
@@ -0,0 +1,636 @@
+1. assumed-type (`TYPE(*)`)
+1. coarray dummy argument
+1. `INTENT(IN) POINTER` argument (15.5.2.7, C.10.4)
+
+Non-F77ish procedures use descriptors to represent two further
+kinds of dummy arguments:
+
+1. assumed-length `CHARACTER(*)`
+1. dummy procedures
+
+F77ish procedures use other means to convey character length and host instance
+links (respectively) for these arguments.
+
+Function results are described by the caller & callee in
+a caller-supplied descriptor when they have any of the following
+characteristics, some which necessitate an explicit interface:
+
+1. deferred-shape array (so `ALLOCATABLE` or `POINTER`)
+1. derived type with any non-constant `LEN` parameter
+   (C795 prohibit assumed lengths)
+1. procedure pointer result (when the interface must be explicit)
+
+Storage for a function call's result is allocated by the caller when
+possible: the result is neither `ALLOCATABLE` nor `POINTER`,
+the shape is scalar or explicit, and the type has `LEN` parameters
+that are constant expressions.
+In other words, the result doesn't require the use of a descriptor
+but can't be returned in registers.
+This allows a function result to be written directly into a local
+variable or temporary when it is safe to treat the variable as if
+it were an additional `INTENT(OUT)` argument.
+(Storage for `CHARACTER` results, assumed or explicit, is always
+allocated by the caller, and the length is always passed so that
+an assumed-length external function will work when eventually
+called from a scope that declares the length that it will use
+(15.5.2.9 (2)).)
+
+Note that the lower bounds of the dimensions of non-`POINTER`
+non-`ALLOCATABLE` dummy argument arrays are determined by the
+callee, not the caller.
+(A Fortran pitfall: declaring `A(0:9)`, passing it to a dummy
+array `D(:)`, and assuming that `LBOUND(D,1)` will be zero
+in the callee.)
+If the declaration of an assumed-shape dummy argument array
+contains an explicit lower bound expression (R819), its value
+needs to be computed by the callee;
+it may be captured and saved in the incoming descriptor
+as long as we assume that argument descriptors can be modified
+by callees.
+Callers should fill in all of the fields of outgoing
+non-`POINTER` non-`ALLOCATABLE` argument
+descriptors with the assumption that the callee will use 1 for
+lower bound values, and callees can rely on them being 1 if
+not modified.
+
+### Copying temporary storage back into argument designators
+
+Except for `VALUE` and known `INTENT(IN)` dummy arguments and array sections
+with vector-valued subscripts (15.5.2.4(21)), temporary storage into
+which effective argument data were compacted for contiguity before the call
+must be redistributed back to its original storage by the caller after
+the return.
+
+In conjunction with saved cosubscript values, a standard descriptor
+suffices to represent a pointer to the original storage into which the
+temporary data should be redistributed.
```

Comment:
Do you mean we would actually create two descriptors for such case inside the caller ? One for the actual argument and one for the dummy and we would do inside the caller :
```
Copy("dummy descriptor", "actual descriptor") // maybe with runtime help for PDTS
Call("dummy descriptor")
Copy("actual descriptor", "dummy descriptor") 
```